pyvo/mivot/viewer/xml_viewer.py

```diff
@@ -0,0 +1,169 @@
+        If all_instances is True, returns a list of all instances matching with @dmtype.
+        Parameters
+        ----------
+        dmtype : str
+            The @dmtype to look for.
+        all : bool, optional
+            If True, returns a list of all instances, otherwise returns the first instance.
+            Default is False.
+        Returns
+        -------
+        Union[~`xml.etree.ElementTree.Element`, List[~`xml.etree.ElementTree.Element`], None]
+            If all_instances is False, returns the instance matching with @dmtype.
+            If all_instances is True, returns a list of all instances matching with @dmtype.
+            If no matching instance is found, returns None.
+        Raises
+        ------
+        MivotElementNotFound
+            If dmtype is not found.
+        """
+        if all_instances is False:
+            if len(XPath.x_path(self._xml_view,
+                                f'.//INSTANCE[@dmtype="{dmtype}"]')) != 0:
+                for ele in XPath.x_path(self._xml_view,
+                                        f'.//INSTANCE[@dmtype="{dmtype}"]'):
+                    if ele is not None:
+                        return ele
+            else:
+                raise MivotElementNotFound(f"Cannot find dmtype {dmtype} in any instances of the VOTable")
+        else:
+            if len(XPath.x_path(self._xml_view,
+                                f'.//INSTANCE[@dmtype="{dmtype}"]')) != 0:
+                ele = []
+                for elem in XPath.x_path(self._xml_view,
+                                         f'.//INSTANCE[@dmtype="{dmtype}"]'):
+                    ele.append(elem)
+                if ele:
+                    return ele
+            else:
+                raise MivotElementNotFound(f"Cannot find dmtype {dmtype} in any instances of the VOTable")
+            return ele
+        return None
+
+    def get_collection_by_role(self, dmrole, all_instances=False):
+        """
+        Return the collection matching with @dmrole.
+        If all_instances is False, returns the first COLLECTION matching with @dmrole.
+        If all_instances is True, returns a list of all COLLECTION matching with @dmrole.
+        Parameters
+        ----------
+        dmrole : str
+            The @dmrole to look for.
+        all_instances : bool, optional
+            If True, returns a list of all COLLECTION, otherwise returns the first COLLECTION.
+            Default is False.
+        Returns
+        -------
+        Union[~`xml.etree.ElementTree.Element`, List[~`xml.etree.ElementTree.Element`], None]
+            If all_instances is False, returns the collection matching with @dmrole.
+            If all_instances is True, returns a list of all collections matching with @dmrole.
+            If no matching collection is found, returns None.
+        Raises
+        ------
+        MivotElementNotFound
+            If dmrole is not found.
+        """
+        if all_instances is False:
```

Comment:
The logic in these 3 methods is very similar (if not the same). Can code duplication be avoided?

Reply:
Duplicated (or similar) code in the 3 element selectors moved to a static method of the `XPath` utility. 
The logic differences between those selectors gets more visible (I hope)